### ISSUE TITLE
Benchmark against level

### ISSUE THREAD

**Issue by vweevers** [Level/abstract-level#4]
Compare:

- [x] `level@7` against `classic-level`. To test the native part (now using encoding options instead of `*asBuffer`). Not expecting a change here.
- [x] `level-mem` + `subleveldown` against `memory-level` + sublevels, with json encoding. To test the JS part. Expecting a slight improvement here, though it might only surface in real-world apps (with GC pressure) rather than a synthetic benchmark.
- [x] Batching on `level-mem` against `memory-level`, as it removes two (or three, with sublevels) iterations of the batch array.

A quick benchmark is enough (of reads and writes). It's just to check that performance is equal or better.

**Comment by vweevers** [comment]
puts on `level-mem` + `subleveldown` versus `memory-level` + sublevels, with json encoding. Win.

![put 1642345597518](https://user-images.githubusercontent.com/3055345/149666032-846bf915-d865-4133-9ed1-fe509b8c82ae.png)

puts on `level-mem` versus `memory-level` versus `memory-level` using strings internally. Double win.

![put 1642346203027](https://user-images.githubusercontent.com/3055345/149666088-3e3b4acb-bcb5-44ff-a279-949d054f46c4.png)

**Comment by vweevers** [comment]
`iterator.next()` on `level-mem` versus `memory-level`, using json and utf8 valueEncodings. No difference.

![iterate 1642346920934](https://user-images.githubusercontent.com/3055345/149666480-26121795-c734-4722-8c5a-86f4fa58f1aa.png)

**Comment by vweevers** [comment]
`iterator.next()` on `level-mem` versus `iterator.nextv(1000)` on `memory-level`. Not a fair benchmark, but the new `nextv()` API is an obvious win.

![iterate 1642347677129](https://user-images.githubusercontent.com/3055345/149666905-8265da65-f184-4130-bd7d-6091b9e112b8.png)

**Comment by vweevers** [comment]
`iterator.next()` on `level` versus `iterator.next()` on `classic-level`. Slower. I reckon that's because I changed the structure of the cache (in short: `[entry, entry, ..]` instead of `[key, value, key, value, ..]`) which should make `nextv()` faster. That'll be difficult to compare fairly.

![iterate 1642373843003](https://user-images.githubusercontent.com/3055345/149681927-461100b1-f5a0-4b95-99e1-76e7beaf9001.png)

**Comment by vweevers** [comment]
Gets on `level-mem` versus `memory-level`. Win.

![get 1643397969100](https://user-images.githubusercontent.com/3055345/151609279-5d5c802b-831f-4a93-ac33-ff6000875019.png)

However, `memory-level` is slower when using a binary valueEncoding. That warrants a closer look.

**Comment by vweevers** [comment]
> However, `memory-level` is slower when using a binary valueEncoding. That warrants a closer look.

It's not due to binary. Happens on any encoding when this code path is triggered:

https://github.com/Level/abstract-level/blob/d711af39de3126ee984d57df396a6d084ebfb748/abstract-level.js#L299-L301

V8 has [a performance issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1204540) with the spread operator when properties are not present. The following "fixes" it:

```js
options.keyEncoding = keyFormat
options.valueEncoding = valueFormat
options = { ...options, keyEncoding: keyFormat, valueEncoding: valueFormat }
```

As does using `Object.assign()` instead of spread:

![get 1643406383068](https://user-images.githubusercontent.com/3055345/151626620-3e939fd4-62ee-4cdb-a054-30d574436700.png)

Could switch to `Object.assign()` but I do still generally prefer the spread operator, for being idiomatic (not being vulnerable to prototype pollution could be another argument but I don't see how that would matter here). Plus, we're only gonna notice the performance regression on `memory-level`, because it's so fast.

**Comment by vweevers** [comment]
The same `get()` performance regression exists on `classic-level`. Using `Object.assign()` would fix it.

![get 1643410282728](https://user-images.githubusercontent.com/3055345/151632568-c1ecd01b-8ebd-4983-8b0b-c89fca6c9198.png)

**Comment by vweevers** [comment]
Quick-and-dirty benchmark of streams, comparing `nextv()` to `next()`. Ref https://github.com/Level/community/issues/70.

Unrelated to `abstract-level`, but it's a win.

```
classic-level | using nextv() | took 1775 ms, 563380 ops/sec
classic-level | using nextv() | took 1577 ms, 634115 ops/sec
classic-level | using nextv() | took 1549 ms, 645578 ops/sec
classic-level | using nextv() | took 1480 ms, 675676 ops/sec
classic-level | using nextv() | took 1572 ms, 636132 ops/sec
                                 avg 1591 ms

level         | using next()  | took 1766 ms, 566251 ops/sec
level         | using next()  | took 1776 ms, 563063 ops/sec
level         | using next()  | took 1737 ms, 575705 ops/sec
level         | using next()  | took 1711 ms, 584454 ops/sec
level         | using next()  | took 1729 ms, 578369 ops/sec
                                 avg 1744 ms
```

**Comment by vweevers** [comment]
Did a better benchmark of streams. This one takes some explaining. In the graph legend below:

- _new-nextv_ is using a `level-read-stream` on a `classic-level` iterator using `nextv(size)`
- _old-next_ is `level().createReadStream()`, i.e. `level-iterator-stream` on a `leveldown` iterator using `next()`
- _new-next_ is using a `level-read-stream` on a `classic-level` iterator using `next()` (a temporary code path for fair benchmarking)
- _new-nextv-tweaked_ uses userland options to make sure that the byte-hwm is more than the expected byte size of a `nextv(size)` array, because otherwise `classic-level` would hit the byte-hwm first and return partially filled arrays
- _old-next-tweaked_ uses userland options to increase both byte-hwm and stream-hwm (manually creating a `level-iterator-stream` so that there's a way to specify both byte-hwm and stream-hwm). In such a way that the byte-hwm is effectively ignored and we can compare the effect of merely increasing stream-hwm.

Where "byte-hwm" is the highWaterMark on the C++ side, measured in bytes. And "stream-hwm" is the highWaterMark of object-mode streams, measured in amount of entries.

That's about half of the explainer needed... In hindsight I wish I didn't do the `abstract-level` and `nextv()` work in parallel. So please allow me to just skip to conclusions (and later document how a user should tweak their options):

- `nextv(size)` is faster than `next()` (compare _new-nextv_ to _old-next_)
- Though the refactorings needed to implement `nextv(size)` make `next()` slightly slower (compare _old-next_ to _new-next_)
- Both the old `next()` and the new `nextv(size)` can be tweaked through options, but `nextv(size)` can't be beaten.
- Bottom line, streams became faster. I won't give any numbers, because there are too many factors.

**TLDR**: we're good. Most importantly, the performance characteristics of streams and iterators did not change, in the sense that an app using smaller or larger values (I used 100 bytes) would be hurt by upgrading to `abstract-level` or `classic-level`. That's because `leveldown` internally already had two highWaterMark mechanisms; `classic-level` merely "hoists" one of them up to streams. So if an app has extremely large values, we will not prefetch more items than before. If an app has small values, we will not prefetch less than before. If an app is not using streams, iterators still do prefetch (as you can see later when I finally push all code).

![stream 1643555843619](https://user-images.githubusercontent.com/3055345/151705875-61423986-4482-4445-bde8-89f8147ebdff.png)

**Comment by vweevers** [comment]
> and later document how a user should tweak their options

Done in https://github.com/Level/classic-level/pull/1
